File: tool.py
#!/usr/bin/env python3
"""
YouTube Courses Tool
Usage: python tool.py add "Titulo" "Duracion" "Descripcion" "Dificultad" "URL" [Categoria]

Example: python tool.py add "Curso de ChatGPT" "2h 15m" "Aprende ChatGPT" "Intermedio" "https://youtube.com/watch?v=ZZq4TpNgnvg"
"""

import sys
import csv
import re
from pathlib import Path
from urllib.parse import urlparse, parse_qs

def get_video_id(url):
    """Extract YouTube video ID from URL"""
    try:
        parsed = urlparse(url)
        if parsed.hostname in ['www.youtube.com', 'youtube.com', 'm.youtube.com']:
            return parse_qs(parsed.query).get('v', [None])[0]
        elif parsed.hostname in ['youtu.be', 'www.youtu.be']:
            return parsed.path[1:]
    except Exception:
        pass
    return None

def escape_html(text):
    """Escape HTML entities"""
    return (text
        .replace('&', '&amp;')
        .replace('<', '&lt;')
        .replace('>', '&gt;')
        .replace('"', '&quot;'))

def add_course(title, duration, description, difficulty, url, category='IA'):
    """Add course to CSV and HTML"""
    base_path = Path(__file__).parent
    csv_path = base_path / 'data' / 'cursos.csv'
    html_path = base_path / 'cursos.html'
    
    video_id = get_video_id(url)
    if not video_id:
        print("Error: Invalid YouTube URL")
        return False
    
    # Add to CSV
    csv_exists = csv_path.exists()
    with open(csv_path, 'a', newline='', encoding='utf-8') as f:
        writer = csv.writer(f)
        if not csv_exists:
            writer.writerow(['titulo', 'categoria', 'duracion', 'descripcion', 'dificultad', 'url'])
        writer.writerow([title, category, duration, description, difficulty, url])
    print(f"[OK] Added to CSV: {title}")
    
    # Add to HTML
    if html_path.exists():
        html_content = html_path.read_text(encoding='utf-8')
        
        html_course = f'''
                    <!-- {escape_html(title)} -->
                    <div class="resource-card" data-category="{category.lower()}">
                        <div class="resource-preview">
                            <img src="https://img.youtube.com/vi/{video_id}/maxresdefault.jpg" alt="{escape_html(title)}">
                        </div>
                        <div class="resource-content">
                            <h3 class="resource-title">{escape_html(title)}</h3>
                            <p class="resource-description">{escape_html(description)}</p>
                            <div class="resource-meta">
                                <span class="resource-duration"><i class="far fa-clock"></i> {duration}</span>
                                <span class="resource-level"><i class="fas fa-signal"></i> {difficulty}</span>
                            </div>
                            <a href="{url}" class="resource-link" target="_blank">Ver Recurso</a>
                        </div>
                    </div>'''
        
        # Insert before closing tags
        html_content = html_content.replace(
            '</div>\n                </div>\n            </div>\n        </section>',
            html_course + '\n                </div>\n            </div>\n        </section>'
        )
        
        html_path.write_text(html_content, encoding='utf-8')
        print(f"[OK] Added to HTML: {title}")
    
    print("[OK] Course added successfully!")
    return True

def list_courses():
    """List all courses"""
    base_path = Path(__file__).parent
    csv_path = base_path / 'data' / 'cursos.csv'
    
    if not csv_path.exists():
        print("No courses found")
        return
    
    with open(csv_path, 'r', encoding='utf-8') as f:
        reader = csv.DictReader(f)
        for i, row in enumerate(reader, 1):
            print(f"{i}. {row['titulo']} | {row['categoria']} | {row['duracion']}")

def main():
    args = sys.argv[1:]
    
    if not args or args[0] in ['help', '-h', '--help']:
        print('''
YouTube Courses Tool
=================

Usage:
  python tool.py add "Titulo" "Duracion" "Descripcion" "Dificultad" "URL" [Categoria]
  python tool.py list

Examples:
  python tool.py add "Curso de ChatGPT" "2h 15m" "Aprende ChatGPT" "Intermedio" "https://youtube.com/watch?v=ZZq4TpNgnvg" "IA"
  python tool.py add "Curso de Midjourney" "1h 30m" "Genera imágenes con IA" "Principiante" "https://youtube.com/watch?v=mmfz73EW60w"
  python tool.py list

Categories: Programación, Web, Ciberseguridad, Datos, Ofimática, Redes, Idiomas, IA
        ''')
        return
    
    if args[0] in ['list', 'ls']:
        list_courses()
        return
    
    if args[0] in ['add', 'a']:
        if len(args) < 6:
            print("Usage: python tool.py add \"Titulo\" \"Duracion\" \"Descripcion\" \"Dificultad\" \"URL\" [Categoria]")
            return
        
        title = args[1]
        duration = args[2]
        description = args[3]
        difficulty = args[4]
        url = args[5]
        category = args[6] if len(args) > 6 else 'IA'
        
        add_course(title, duration, description, difficulty, url, category)
        return
    
    print("Unknown command. Run: python tool.py help")

if __name__ == '__main__':
    main()

File: youtube-courses-tool/tool.js
#!/usr/bin/env node

const https = require('https');
const fs = require('fs');
const path = require('path');
const { URL } = require('url');

function getYouTubeVideoId(url) {
    try {
        const urlObj = new URL(url);
        if (urlObj.hostname.includes('youtube.com') || urlObj.hostname.includes('youtu.be')) {
            return urlObj.searchParams.get('v') || urlObj.pathname.slice(1);
        }
    } catch (e) {}
    return null;
}

function extractMetadata(html) {
    const result = {
        title: '',
        duration: '',
        description: ''
    };

    const titleMatch = html.match(/<title>([^<]+)<\/title>/i);
    if (titleMatch) {
        result.title = titleMatch[1].replace(' - YouTube', '').trim();
    }

    const durationMatch = html.match(/lengthSeconds":"(\d+)"/);
    if (durationMatch) {
        const seconds = parseInt(durationMatch[1]);
        const hours = Math.floor(seconds / 3600);
        const minutes = Math.floor((seconds % 3600) / 60);
        const secs = seconds % 60;
        
        if (hours > 0) {
            result.duration = `${hours}h ${minutes}m`;
        } else if (minutes > 0) {
            result.duration = `${minutes}m ${secs}s`;
        } else {
            result.duration = `${secs}s`;
        }
    }

    const shortDescMatch = html.match(/"shortDescription":"([^"]+)"/);
    if (shortDescMatch) {
        result.description = shortDescMatch[1].substring(0, 200);
    }

    return result;
}

function fetchYouTubeVideo(url) {
    return new Promise((resolve, reject) => {
        const urlObj = new URL(url);
        const videoId = getYouTubeVideoId(url);
        
        if (!videoId) {
            reject(new Error('Invalid YouTube URL'));
            return;
        }

        const options = {
            hostname: 'www.youtube.com',
            path: `/watch?v=${videoId}`,
            method: 'GET',
            headers: {
                'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
            }
        };

        const req = https.request(options, (res) => {
            let data = '';
            res.on('data', (chunk) => data += chunk);
            res.on('end', () => {
                resolve(extractMetadata(data));
            });
        });

        req.on('error', reject);
        req.end();
    });
}

function addCourseToCSV(course) {
    const csvPath = path.join(__dirname, 'data', 'cursos.csv');
    const fileContent = fs.readFileSync(csvPath, 'utf-8');
    const rows = fileContent.split('\n');
    
    const newRow = `"${course.title}","IA","${course.duration}","${course.description}","Intermedio","${course.url}"`;
    
    const header = rows[0];
    rows.splice(1, 0, newRow);
    
    fs.writeFileSync(csvPath, rows.join('\n'), 'utf-8');
    console.log('Course added to CSV:', course.title);
}

function addCourseToHTML(course) {
    const htmlPath = path.join(__dirname, 'cursos.html');
    let content = fs.readFileSync(htmlPath, 'utf-8');
    
    const courseTemplate = `
                    <!-- Nuevo Curso -->
                    <div class="resource-card" data-category="ia">
                        <div class="resource-preview">
                            <img src="https://img.youtube.com/vi/${getYouTubeVideoId(course.url)}/maxresdefault.jpg" alt="${course.title}">
                        </div>
                        <div class="resource-content">
                            <h3 class="resource-title">${course.title}</h3>
                            <p class="resource-description">${course.description}</p>
                            <div class="resource-meta">
                                <span class="resource-duration"><i class="far fa-clock"></i> ${course.duration}</span>
                                <span class="resource-level"><i class="fas fa-signal"></i> Intermedio</span>
                            </div>
                            <a href="${course.url}" class="resource-link" target="_blank">Ver Recurso</a>
                        </div>
                    </div>`;
    
    content = content.replace('</div>\n                </div>\n            </div>\n        </section>', courseTemplate + '\n                </div>\n            </div>\n        </section>');
    
    fs.writeFileSync(htmlPath, content, 'utf-8');
    console.log('Course added to HTML:', course.title);
}

async function main() {
    const args = process.argv.slice(2);
    const command = args[0];
    
    if (command === 'add') {
        const url = args[1];
        if (!url) {
            console.error('Usage: node tool.js add <youtube-url>');
            process.exit(1);
        }
        
        console.log('Fetching metadata from:', url);
        const metadata = await fetchYouTubeVideo(url);
        
        const course = {
            title: metadata.title,
            duration: metadata.duration,
            description: metadata.description,
            url: url
        };
        
        console.log('\n--- Course Info ---');
        console.log('Title:', course.title);
        console.log('Duration:', course.duration);
        console.log('Description:', course.description);
        
        addCourseToCSV(course);
        addCourseToHTML(course);
        
    } else if (command === 'help') {
        console.log(`
YouTube Courses Tool
=============

Usage:
  node tool.js add <youtube-url>   Add a new course from YouTube URL
  node tool.js help            Show this help message

Examples:
  node tool.js add https://www.youtube.com/watch?v=ZZq4TpNgnvg
        `);
    } else {
        console.log('Unknown command. Run: node tool.js help');
    }
}

main().catch(console.error);